src/main/kotlin/org/axonframework/intellij/ide/plugin/resolving/MessageHandlerResolver.kt

```diff
@@ -0,0 +1,85 @@
+package org.axonframework.intellij.ide.plugin.resolving
+
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiClassObjectAccessExpression
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiLiteralExpression
+import com.intellij.psi.PsiMethod
+import com.intellij.psi.PsiType
+import com.intellij.psi.search.searches.AnnotatedElementsSearch
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import org.axonframework.intellij.ide.plugin.api.MessageHandler
+import org.axonframework.intellij.ide.plugin.api.MessageHandlerType
+import org.axonframework.intellij.ide.plugin.util.areAssignable
+import org.axonframework.intellij.ide.plugin.util.axonScope
+import org.axonframework.intellij.ide.plugin.util.comparePsiElementsBasedOnDisplayName
+import org.axonframework.intellij.ide.plugin.util.toQualifiedName
+
+class MessageHandlerResolver(private val project: Project) {
+    private val handlerCache = CachedValuesManager.getManager(project).createCachedValue() {
+        CachedValueProvider.Result.create(executeFindMessageHandlers(), PsiModificationTracker.MODIFICATION_COUNT)
+    }
+
+    fun findHandlersForType(qualifiedName: String): List<MessageHandler> {
+        return handlerCache.value
+                .filter { areAssignable(project, it.payloadType, qualifiedName) }
+                .filter { it.element.isValid }
+    }
+
+    fun findAllHandlers() = handlerCache.value
+
+    fun findHandlerByElement(psiElement: PsiElement): MessageHandler? {
+        return handlerCache.value.firstOrNull { it.element == psiElement }
+    }
+
+    private fun executeFindMessageHandlers(): List<MessageHandler> {
+        val annotatedMethods = MessageHandlerType.findAnnotationsForProject(project).flatMap { annotationClass ->
+            AnnotatedElementsSearch.searchPsiMethods(annotationClass, project.axonScope()).findAll()
+        }
+
+        val handlers = annotatedMethods.mapNotNull {
+            val type = resolveType(it)?.toQualifiedName() ?: return@mapNotNull null
+            val handlerType = resolveHandlerType(it) ?: return@mapNotNull null
+            MessageHandler(it, handlerType, type, resolveProjectionName(it))
```

Comment:
So, for event handlers the `resolveProjectionName` makes sense, but not for all messages.

Added, the assumption right now is that an event handler always belongs in/to a projection, while an event handler can be used for a multitude of different tasks.
So, renaming the method to something like `resolveProcesingGroup` sounds feasible to me.